### ISSUE TITLE
`tonumber(tostring(42.0))==nil` on OS with non-English locale

### ISSUE THREAD

**Issue by Egor-Skriptunoff** [fengari-lua/fengari#136]
On vanilla Lua, scripts run in "C" locale (and can switch to another locale if they want).
On Fengari, Lua scripts run in OS-specific-locale and can't switch to "C" locale because `os.setlocale()` is not implemented.

----

**Vanilla Lua** running under Russian Windows (decimal point symbol in Russian locale is comma):
```
   Lua 5.4.0  Copyright (C) 1994-2018 Lua.org, PUC-Rio

   > print(tonumber(tostring(42.0)), tonumber("42.0"), tonumber("42,0"))
   42.0    42.0    nil

   > os.setlocale""
   Russian_Russia.1251

   > print(tonumber(tostring(42.0)), tonumber("42.0"), tonumber("42,0"))
   42,0    42,0    42,0

   > os.setlocale"C"
   C

   > print(tonumber(tostring(42.0)), tonumber("42.0"), tonumber("42,0"))
   42.0    42.0    nil

```
----
**Fengari** inside US-English FireFox running under Russian Windows:
```
   > print(tonumber(tostring(42.0)), tonumber("42.0"), tonumber("42,0"))
   nil	42,0	nil

   > os.setlocale"C"
   [string "stdin"]:1: attempt to call a nil value (field 'setlocale')
   stack traceback:
      [string "stdin"]:1: in main chunk
      [JS]: in function 'xpcall'
      https://fengari.io/static/lua/web-cli.lua?v=1:81: in upvalue 'doREPL'
      https://fengari.io/static/lua/web-cli.lua?v=1:106: in function <https://fengari.io/static/lua/web-cli.lua?v=1:98>

```
Could "C" locale be hardcoded (instead of OS-specific-locale) to make `tostring()` and `tonumber()` more predictable and consistent?

**Comment by daurnimator** [comment]
fengari can only use what JavaScript exposes. In the case of printing a number, that is https://github.com/fengari-lua/fengari/blob/c3c170329de1ede2bc970207f6be1315634b099d/src/luaconf.js#L153-L155

As far as I know, Javascript has no way to change the locale used?

**Comment by daurnimator** [comment]
Oh I see, I think the issue is that the regex in `lua_str2number` here: https://github.com/fengari-lua/fengari/blob/c3c170329de1ede2bc970207f6be1315634b099d/src/lobject.js#L488
Only allows '.' and not the locale-specific decimal point

**Comment by Egor-Skriptunoff** [comment]
>  const lua_getlocaledecpoint = function() { 
>      return (1.1).toLocaleString().charCodeAt(1); 
>  }; 

BTW, `1.5` might be safer than `1.1`

**Comment by Egor-Skriptunoff** [comment]
>  const lua_number2str = function(n) { 
>      return String(Number(n.toPrecision(14))); /* should match behaviour of LUA_NUMBER_FMT */ 
>  }; 

Could the following formula be used to mimic vanilla Lua behavior (Lua uses "C" locale, which is US-English AFAIK)?
`Number(n.toPrecision(14)).toLocaleString('en-US');`

**Comment by daurnimator** [comment]
> Lua uses "C" locale

Lua uses the locale of the current process, whatever that may be (and it's dependant on what the host program is doing)

On unix, new processes are always started with the C locale IIRC, so the lua interpreter will use the C locale by default, as it doesn't call `setlocale` as part of startup.

**Comment by Egor-Skriptunoff** [comment]
> > Lua uses "C" locale
> 
> Lua uses the locale of the current process, whatever that may be (and it's dependant on what the host program is doing)
> 
> On unix, new processes are always started with the C locale IIRC, so the lua interpreter will use the C locale by default, as it doesn't call `setlocale` as part of startup.

That's correct.
On Windows Lua starts with C locale too (see my example in the first post).
IMO, this behavior is worth propagating to Fengari.
I want Lua script in webpage to be independent of current OS locale.
I don't like locale-related surprises :-)

**Comment by daurnimator** [comment]
> BTW, `str:lower()` and `str:upper()` function do not respect OS locale.

Good catch; though I think my answer is the opposite way to what you want: we should be using [`toLocaleUpperCase`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleUpperCase) rather than the current speed hack... https://github.com/fengari-lua/fengari/blob/c3c170329de1ede2bc970207f6be1315634b099d/src/lstrlib.js#L682

**Comment by Egor-Skriptunoff** [comment]
> > BTW, `str:lower()` and `str:upper()` function do not respect OS locale.
> 
> Good catch; though I think my answer is the opposite way to what you want: we should be using [`toLocaleUpperCase`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleUpperCase) rather than the current speed hack...

OK.
That makes sense.

But there is a difficulty here: on Windows there are two encodings in use: 1-byte (windows-1251) and UTF-8.
Probably, the simplest solution you could implement would respect only native Windows encoding (1-byte).
But the most of usage is UTF-8 (for example, the demo on `fengari.io` uses UTF-8 for strings).

So, fengari-powered web sites may have problems with Lua functions `upper()` and `lower()` on non-English Windows hosts.

**Comment by daurnimator** [comment]
fengari uses 8-bit clean strings the same as lua does. Even on windows lua just uses 1-byte case functions: fengari will do the same.

**Comment by Egor-Skriptunoff** [comment]
It is now clear that locale support in `lower()` and `upper()` would be practically useless :-)

**Comment by Egor-Skriptunoff** [comment]
Not only locale support in `lower()` and `upper()` would be useless, it actually might be dangerous!


Introducing locale support for `utf8.lower()` and `utf8.upper()` would be catastrophic on Turkish locale.
Fengari-powered Lua function `string.lower()` would convert the letter "I" ("\x49") to "U+0131 LATIN SMALL LETTER DOTLESS I" which is "\xFD" on windows-1254 (native codepage on Turkish Windows).
So, valid UTF-8 string containing pure English text would be converted to invalid UTF-8 string by applying `string.lower()`.
And web developer doesn't have a chance to avoid this problem by explicitly switching from OS locale to "C" locale because `os.setlocale()` is not implemented in Fengari.

> "the intention of fengari is to follow Lua as close as possible"

I like this intention.

As I see, there are three ways to correctly "follow Lua as close as possible":
- emulate `os.setlocale()` (every locale-sensitive call in JS code should specify locale identifier selected by user or use `navigator.language` while locale is not yet selected) to follow embedded Lua
- or permanently stick to "C" locale (by explicitly specifying 'en-US' on every locale-sensitive call in JS code) to follow standalone Lua with `os.setlocale` deprecated
- or draw big red banner "Not for Turkish users" on every Fengari-based website :-)

If a Fengari-powered website is targeted to the visitors from the whole World (including Turkey), then web developer have to start Lua script with `os.setlocale("C")`.
May be, "sticking to C locale" behavior is not that bad?  :-)

**Comment by daurnimator** [comment]
>     * emulate `os.setlocale()` (every locale-sensitive call in JS code should specify locale identifier 
selected by user or use `navigator.language` while locale is not yet selected) to follow embedded Lua

`os.setlocale` sets the default locale for the *process*. There is no way to do that in JS, so this isn't an option

> or permanently stick to "C" locale (by explicitly specifying 'en-US' on every locale-sensitive call in JS code) to follow standalone Lua with os.setlocale deprecated

This is probably most reasonable..... though I think people expect that passing a string obtained from `1.0+""` to lua would be parseable with `tonumber`. I'm not sure this would be preserved in this case?

**Comment by Egor-Skriptunoff** [comment]
> > or permanently stick to "C" locale (by explicitly specifying 'en-US' on every locale-sensitive call in JS code) to follow standalone Lua with os.setlocale deprecated
> 
> This is probably most reasonable..... though I think people expect that passing a string obtained from `1.0+""` to lua would be parseable with `tonumber`. I'm not sure this would be preserved in this case?

As a workaround, `tonumber` may fallback to OS locale if "C" locale fails to read the number.
This probably would not break any code written for Lua 5.3.

All other locale-sensitive Lua operations (`tostring`, `lower`, `upper`, patterns `%a`, `%l`, `%u` and string comparison `"abc" < "def"`) should act according to "C" locale.

**Comment by Egor-Skriptunoff** [comment]
It's good that a decision has been made to stick to "C" locale.
But this decision is currently hidden from Lua scripts.
A Lua script or library might want to invoke `os.setlocale()` to determine the current locale (without intention to change the locale).

IMO, `os.setlocale()` should be implemented (as a stub) to explicitly emulate a platform which doesn't support other locales except "C".
```lua
   os.setlocale()                   should return "C"
   os.setlocale("C")                should return "C"
   os.setlocale("")                 should return "C"
   os.setlocale("any_other_locale") should return nil
```

**Comment by daurnimator** [comment]
@Egor-Skriptunoff good idea. fixed via #151
